text/0000-nested-publish.md

```diff
@@ -0,0 +1,359 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow published Cargo packages to depend on other packages stored within themselves, as is currently possible in unpublished packages.
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
```

Comment:
Another use case is that this provides another way for us to break dependency cycles that involve `dev-dependencies`.

Currently, the solution involves dropping the dependency on publish (by not specifying a `version`).  This lacked discovery so by default `cargo add` does it for *all* path dev-dependencies. This negatively impacts crater because it means that any packages with dev-dependency cycles or where `cargo add` was used to add a path dev-dependency, we lose out on a lot of testing with crater.

With this feature, we can instead nest the path dev-dependency.